### ISSUE TITLE
Microsoft Office apps can't write on an S3 mount when mounted with nfsmount

### ISSUE THREAD

**Issue by alexbet** [rclone/rclone#7973]
#### What is the problem you are having with rclone?
I started using the nfsmount option to mount Amazon S3 drives because it is very problematic to give the proper permisson for the kernel extensions for MacFuse. At first, I thought that the nfs mount worked perfectly. It does, but then I realized that Microsoft Office apps Word, Excel, Powerpoint, can NOT write on the mounted drive. All other apps on my Mac can do that, just not the Microsoft apps, which is very weird. I tried playing with these options

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by kapitainsky** [comment]
Remove FUSE, install [FUSE-T](https://www.fuse-t.org/) and use `rclone mount`. It it kext-less and behind the hood it uses nfs as `rclone nfsmount` but from my experience is more bugs free. At least for me no problem with opening and writing MS Office docs.

**Comment by alexbet** [comment]
> Remove MacFUSE (if you have it), install [FUSE-T](https://www.fuse-t.org/) (available via brew) and use `rclone mount`. It is kext-less and under the hood it uses nfs, the same as `rclone nfsmount` but from my experience is more stable and compliant (it uses macOS nfs server and not some home brewed implementation). At least for me no problem with opening and writing MS Office docs.

Worked like a charm! Thanks @kapitainsky!

**Comment by ncw** [comment]
Thanks for solving @alexbet 's problem @kapitainsky 

I would still like to to figure out what is wrong with the NFS server though. There is nothing suspicious in the log - no errors or anything, which makes me think that rclone has set the wrong permissions.

I have been working on the NFS server recently though and I fixed several bugs which might be relevant - can you try this branch please @alexbet 

[v1.68.0-beta.8062.a5e84274e.fix-nfs-handles](https://beta.rclone.org/branch/fix-nfs-handles/v1.68.0-beta.8062.a5e84274e.fix-nfs-handles/) on branch [fix-nfs-handles](https://github.com/rclone/rclone/tree/fix-nfs-handles) - you will find some osx binaries in there - I think [this](https://beta.rclone.org/branch/fix-nfs-handles/v1.68.0-beta.8062.a5e84274e.fix-nfs-handles/rclone-v1.68.0-beta.8062.a5e84274e.fix-nfs-handles-osx-arm64.zip) should be the right one.

**Comment by kapitainsky** [comment]
> [v1.68.0-beta.8062.a5e84274e.fix-nfs-handles](https://beta.rclone.org/branch/fix-nfs-handles/v1.68.0-beta.8062.a5e84274e.fix-nfs-handles/) on branch [fix-nfs-handles](https://github.com/rclone/rclone/tree/fix-nfs-handles) - you will find some osx binaries in there - I think [this](https://beta.rclone.org/branch/fix-nfs-handles/v1.68.0-beta.8062.a5e84274e.fix-nfs-handles/rclone-v1.68.0-beta.8062.a5e84274e.fix-nfs-handles-osx-arm64.zip) should be the right one.

This rclone version `rclone nfsmount` works fine with MS Office files - the same as FUSE-T

**Comment by alexbet** [comment]
> [v1.68.0-beta.8062.a5e84274e.fix-nfs-handles](https://beta.rclone.org/branch/fix-nfs-handles/v1.68.0-beta.8062.a5e84274e.fix-nfs-handles/) on branch [fix-nfs-handles](https://github.com/rclone/rclone/tree/fix-nfs-handles) - you will find some osx binaries in there - I think [this](https://beta.rclone.org/branch/fix-nfs-handles/v1.68.0-beta.8062.a5e84274e.fix-nfs-handles/rclone-v1.68.0-beta.8062.a5e84274e.fix-nfs-handles-osx-arm64.zip) should be the right one.

Unfortunately, this does not work for me. I tried opening the existing test.docx file inside the Alex Test folder, but. it opened read only. I also created a new file and tried to save to the same folder as test2.docx, and again it wouldn't let me save it.

Here is the log file

[Terminal Saved Output.txt](https://github.com/user-attachments/files/16432267/Terminal.Saved.Output.txt)

**Comment by kapitainsky** [comment]
> Unfortunately, this does not work for me.

Interesting... can you check your S3 drive? Are there any files named like `.nfs12345.123`

**Comment by alexbet** [comment]
> > Unfortunately, this does not work for me.
> 
> Interesting... can you check your S3 drive? Are there any files named like `.nfs12345.123`

No, I can't find any files like that. Here are all the files, including the deleted ones that are in this folder. BTW, the "testing.rtf" file saves and edits fine with TextEdit.

<img width="586" alt="Screenshot 2024-07-30 at 4 03 17 PM" src="https://github.com/user-attachments/assets/6133cc2f-c216-40a5-843d-3deecd0991f4">

**Comment by alexbet** [comment]
BTW is there an RClone version for Mac Intel chipsets on [v1.68.0-beta.8062.a5e84274e.fix-nfs-handles](https://beta.rclone.org/branch/fix-nfs-handles/v1.68.0-beta.8062.a5e84274e.fix-nfs-handles/) on branch [fix-nfs-handles](https://github.com/rclone/rclone/tree/fix-nfs-handles)? I only see AMD and ARM. I have a computer with an Intel chipset in my office that I can test on, othwerwise I need to do this on one of our employees' computers. Or maybe the AMD one works for Intel?

**Comment by ncw** [comment]
Thanks for testing.The amd binary one will work on Intel. Look forward to the results of your testing.

**Comment by alexbet** [comment]
@ncw , @kapitainsky , OK, so I tested on the Intel computer that I have here for testing, and I tested on a an empty S3 folder so the log file is hopefully easier to read for you. Something weird happened while testing this time.  Here is what happened...

1) I created a new file "test.docx" in MS Word, and tried to save it to S3. It looked like it is saving it, I got a spinning ball for about 20-30 seconds, and it seemed like it has been saved, but Word crashed. 
2) I reopened Word and tried to open the file from the drive, and it opened, but in read only mode. Also, the file didn't have the text that I wrote in it, it was blank. I tried to edit the file, but couldn't because Word said that it was read only.
3) I created a new file "test 2.docx" and tried to save it to the the drive again. Again the same thing happened, it initially saved the file after a while, but then Word switched to read only mode, and there was no content in it, it was empty.

Here is the log file from all of this:
[Terminal Saved Output.txt](https://github.com/user-attachments/files/16433963/Terminal.Saved.Output.txt)

When I finished the session and unmounted the drive, I went to the S3 console from my browser and downloaded the two files just to see what was written. They are 0 (zero) kilobytes in size, and they have no content, they are completely blank. Very weird what happens.

**Comment by ncw** [comment]
@alexbet I think Word crashing is why the file ended up empty.

So we need to work out why Word crashed.

There are no errors or anything unusual in the log you provided.

Are there any errors in the macOS kernel logs? That is probably where NFS errors end up.

BTW I merged those NFS fixes into [the latest beta](https://beta.rclone.org/) now.

**Comment by alexbet** [comment]
@ncw Word crashed the first time I tried it. The second time I tried it, Word didn't crash. just locked up for about 20-30 seconds, and then resumed. The result was the same both times. 

I can test more though. Where do I find the MacOS kernel logs?

**Comment by kapitainsky** [comment]
I can confirm that the latest rclone beta `rclone nfsmount` and MS Word do some strange things... No crashes for me but I can not save any document. And all Word documents when opened are reported as "read-only". Interestingly all works fine when I edit the same files with Libre Office.

`rclone mount` and FUSE-T works perfectly fine.

**Comment by ncw** [comment]
Can you see if you can create files with ~ and $ in using nfsmount - maybe that is the problem?

**Comment by kapitainsky** [comment]
> Can you see if you can create files with ~ and $ in using nfsmount - maybe that is the problem?

Yes and it is even more bizarre... With Excel sidecar is created and file is open normally:) but saving fails too.

What I noticed now:

nfsmount - why root????:
```
$ ll
total 69
-rw-r--r--  1 root  wheel    20K Aug 19 19:30 Manage My Money1.xlsx
-rw-r--r--  1 root  wheel    13K Aug 19 19:52 test.docx
```

mount with FUSE-T:
```
$ ll
total 69
-rw-r--r--  1 kptsky  staff    20K Aug 19 19:30 Manage My Money1.xlsx
-rw-r--r--  1 kptsky  staff    13K Aug 19 19:52 test.docx
```

**Comment by ncw** [comment]
Ah, I suspect that is the real problem, the root user thing.

Can you try setting this up manually and putting `--uid XXX` and `--gid XXX` options on the `rclone serve nfs`?

Or alternatively we can tell the nfs mount to squash the users into the current userid - not sure how to do this on macOS but you squash the userids

**Comment by kapitainsky** [comment]
> not sure how to do this on macOS but you squash the userids

But are you using macOS NFS server? Then it should be the same as on linux - `all_squash` when exporting filesystem.

**Comment by ncw** [comment]
> Unfortunately it does not have any effect. Command I used:
> 
> ```
> $ id -u
> 501
> 
> $ id -g
> 20
> 
> $ rclone-beta nfsmount iDrive:test-uk ~/Temp/mount --vfs-cache-mode=full -vv --uid 501 --gid 20
> ```

And when you did `ls -l` everything was still owned by root? If so that is a bug in nfs mount which shouldn't be hard to fix.
 
> And indeed it might be the root cause of this issue. With T-FUSE mount has read/write permissions for my current user when with `nfsmount` read/write is only for `system`

OK

> On the side note it would also explain why Libre Office works and not MS Office. MS Office is fully sandboxed (think Snap on Linux) and its access is heavily restricted. Libre Office is not:) and can do much more.

:+1:

> > not sure how to do this on macOS but you squash the userids
> 
> But are you using macOS NFS server? Then it should be the same as on linux - `all_squash` when exporting filesystem.

I think I may have confused myself here. The server is responsible for the squash isn't it, not the client and that is bourne out by there not being any squash options in `man mount` for nfs. The server here is `rclone serve nfs` not the macOS NFS server.

It looks like I need to implement this for the nfs server to be able to report uids and gids - it doesn't look too hard

https://github.com/willscott/go-nfs/blob/91bc38957cc95bcbbcc6f41c9bcf72dc8f6f47e0/README.md?plain=1#L86-L91

**Comment by kapitainsky** [comment]
> And when you did `ls -l` everything was still owned by root? If so that is a bug in nfs mount which shouldn't be hard to fix.

Yeap

```
$ rclone-beta nfsmount iDrive:test-uk ~/Temp/mount --vfs-cache-mode=full -vv --uid 501 --gid 20
2024/08/20 17:34:40 DEBUG : rclone: Version "v1.68.0-beta.8236.be448c9e1" starting with parameters ["rclone-beta" "nfsmount" "iDrive:test-uk" "/Users/kptsky/Temp/mount" "--vfs-cache-mode=full" "-vv" "--uid" "501" "--gid" "20"]
...

$ ll ~/Temp/mount 
total 69
-rw-r--r--  1 root  wheel    20K Aug 19 19:30 Manage My Money1.xlsx
-rw-r--r--  1 root  wheel    13K Aug 19 19:52 test.docx
```

**Comment by ncw** [comment]
Ok give this a go

[v1.68.0-beta.8195.f22f86ea5.fix-7973-nfs-server-ids](https://beta.rclone.org/branch/fix-7973-nfs-server-ids/v1.68.0-beta.8195.f22f86ea5.fix-7973-nfs-server-ids/) on branch [fix-7973-nfs-server-ids](https://github.com/rclone/rclone/tree/fix-7973-nfs-server-ids) (uploaded in 15-30 mins)

It should export the UID and GID set by the --uid and --gid flags on `rclone serve nfs` which should in turn be set to your user id and group id by default so you shouldn't need to set them.

Hopefully this will fix the problem!

> What is interesting is that I can create/edit files in this mount from terminal. I can also manage files using macOS file manager - copy new file, delete etc.

I think the NFS server is supposed to check the permissions, and rclone just says, yes OK!

> I think MS Office problem is just when macOS hardening for sandboxed apps kicks in as it does not like mount point permissions:

I think you are right - something is reading the uid/gid/permissions that rclone is exporting and saying no.

Hopefully the above will work. I tested it on Linux and the NFS client was being rather flaky - it is possible the new code is causing that flakiness - not sure.

**Comment by kapitainsky** [comment]
I have tried to build it but:
```
$ go build -tags cmount
...
# github.com/rclone/rclone/cmd/serve/nfs
cmd/serve/nfs/filesystem.go:48:10: cannot use math.MaxUint64 (untyped int constant 18446744073709551615) as int32 value in struct literal (overflows)
```

**Comment by ncw** [comment]
Alas, I messed up the cross platform bits! Here is a new beta which should compile on all architectures that `rclone serve nfs` supports (I tested the previous one compiled on macOS then made a last minute change and forgot to test it again :-( )

[v1.68.0-beta.8195.877f8d586.fix-7973-nfs-server-ids](https://beta.rclone.org/branch/fix-7973-nfs-server-ids/v1.68.0-beta.8195.877f8d586.fix-7973-nfs-server-ids/) on branch [fix-7973-nfs-server-ids](https://github.com/rclone/rclone/tree/fix-7973-nfs-server-ids) (uploaded in 15-30 mins)

**Comment by kapitainsky** [comment]
[v1.68.0-beta.8195.877f8d586.fix-7973-nfs-server-ids](https://beta.rclone.org/branch/fix-7973-nfs-server-ids/v1.68.0-beta.8195.877f8d586.fix-7973-nfs-server-ids/) fixes mount and content ownership/permissions. Now without any uid/gid all is the same as with FUSE-T.

It does NOT change a bit behaviour of MS Office. I did a bit of Google and seems that issue is quite common (on various platforms) for example:

https://www.truenas.com/community/threads/word-for-mac-on-nfs-v4-share-thinks-documents-are-read-only.112183/
https://forums.unraid.net/topic/143944-word-for-mac-on-nfs-share-thinks-documents-are-read-only/

Common advice is to use NFS3 (which I think is exactly what this golang implementation is) but also to use various locking controlling flags like `nolock` when mounting NFS.

**Comment by ncw** [comment]
Interesting. I saw a mention of locking on an issue on the nfs library also https://github.com/willscott/go-nfs/issues/54 which is describing the same issue.

We implement nfsv3 so we should be able to use `nolock` or `locallocks` .

Can you try mounting with that? You should able to add `-o nolocks` to the nfsmount command line.

The lib has a way of implementing locks also so I could add that into rclone which might be preferable if nolocks works.

I haven't tested it on Windows. I don't know how to use nfs on Windows, I'm under the impression that you need pay for software to do it.

**Comment by ncw** [comment]
> Unfortunately no change. The best I got is exactly like in [willscott/go-nfs#54](https://github.com/willscott/go-nfs/issues/54) - exactly the same results with `locallocks`.

@kapitainsky Oh well... 

> I have tried multiple flags using various ideas from Internet - nothing worked. I know that flags are passed as results differ but nothing worked for MS Office apps.

:+1:

Did you try `nolock` too? It sounds like you did.

> I remember similar issued from early days of FUSE-T, including uid/gid problem and locking. But by now it all works. Shame it is not open source so we can't peep into its source code:(

We can peek into the nfs mount options used though I suspect. In linux you'd see these in `/proc/mounts` and you may see them when running the `mount` command too which should work on macOS.

**Comment by kapitainsky** [comment]
> Did you try `nolock` too? It sounds like you did.

sure I did, `nolock` and even just in case 'nolocks':)

> We can peek into the nfs mount options used though I suspect. In linux you'd see these in /proc/mounts and you may see them when running the mount command too which should work on macOS.

My understanding is that FUSE-T uses NFS4 server so locking options are controlled when exporting fs - unlike NFS3 where they can be set by client

**Comment by ncw** [comment]
It would be worth adding `nodev` and `nosuid` options and see if that makes any difference. I'm not hopeful though.

What we need to do is find out what is causing the problem in Word. This is easier said than done. 

What we can do is trace the system calls that word executes using `dtruss` (see this [stack overflow post](https://stackoverflow.com/questions/31045575/how-to-trace-system-calls-of-a-program-in-mac-os-x) for some hints).

I'd like to see exactly what it is trying to do. That may be very difficult to interpret though and it is likely the macOS security will get in the way as these are sandboxed apps.

**Comment by kapitainsky** [comment]
> It would be worth adding `nodev` and `nosuid` options and see if that makes any difference. I'm not hopeful though.

I did try this path - no difference.

> What we can do is trace the system calls that word executes using dtruss (see this [stack overflow post](https://stackoverflow.com/questions/31045575/how-to-trace-system-calls-of-a-program-in-mac-os-x) for some hints).

I will try to produce something this weekend. But I am not sure it will lead to success.

**Comment by ncw** [comment]
Can you try this please?

[v1.68.0-beta.8196.4cd151085.fix-7973-nfs-server-ids](https://beta.rclone.org/branch/fix-7973-nfs-server-ids/v1.68.0-beta.8196.4cd151085.fix-7973-nfs-server-ids/) on branch [fix-7973-nfs-server-ids](https://github.com/rclone/rclone/tree/fix-7973-nfs-server-ids) (uploaded in 15-30 mins)

It has a patched nfs server which ignores the exclusive lock demand which might solve the problem! If it does I'll work to get something like this upstream.

**Comment by kapitainsky** [comment]
I have double checked if running this latest fix and even compiled from source. I was so suspicious as results are exactly the same as before including trying various nfs_mount flags. Like nothing has changed.

**Comment by ncw** [comment]
Disappointing! But I may have messed up the NFS server library patch....

Could you attach a `-vv` log with the output from `rclone serve nfs` when it fails please?

**Comment by ncw** [comment]
Thank you for the logs.

The fundamental problem is this I think

    2024/08/27 17:43:59 ERROR : nfs: failing create to indicate lack of support for 'exclusive' mode.

The author of the nfs library has had a go at implementing this feature, and I've included this in this build

[v1.68.0-beta.8252.ff8ce1bef.fix-7973-nfs-server-ids](https://beta.rclone.org/branch/fix-7973-nfs-server-ids/v1.68.0-beta.8252.ff8ce1bef.fix-7973-nfs-server-ids/) on branch [fix-7973-nfs-server-ids](https://github.com/rclone/rclone/tree/fix-7973-nfs-server-ids) (uploaded in 15-30 mins)

Can you have a go with this and produce the same rclone log please?

The fuse-t log is interesting but unfortunately it is just too different in operation to help diagnose this bug I think.

**Comment by kapitainsky** [comment]
No change in NFS and MS Office apps behaviour. The same test as earlier with the latest fixes.

```
$ git rev-parse --short HEAD
ff8ce1bef
```

[fix-7973-nfs-server-ids-ff8ce1bef.log](https://github.com/user-attachments/files/16785278/fix-7973-nfs-server-ids-ff8ce1bef.log)

**Comment by ncw** [comment]
Thanks @kapitainsky I've been through those logs very carefully.  I can see the new code is working but it hasn't fixed the problem :-(

Maybe we need to take a step back

> I attempt to open this Word document. It is opened but with message
> that it is read-only and I should create duplicate in order to edit so
> I try to save it as "test1.docx". It fails.

Maybe we should try to figure out why word thinks it is read only? What permissions do you get if you do `ls -l /Users/kptsky/Temp/mount`

How do those compare with the permissions that fuse-t gives? I'd be very interested to see

    ls -ld /Users/kptsky/Temp/mount
    ls -l /Users/kptsky/Temp/mount 
    stat /Users/kptsky/Temp/mount
    stat /Users/kptsky/Temp/mount/test.docx

for both rclone nfsmount and fuse-t to see if we've missed something with the unix permissions.

Maybe you could play with

      --dir-perms FileMode                     Directory permissions (default 777)
      --file-perms FileMode                    File permissions (default 666)
      --umask FileMode                         Override the permission bits set 
      --gid uint32                             Override the gid field set by the filesystem (not supported on Windows) (default 1000)
      --uid uint32                             Override the uid field set by the filesystem (not supported on Windows) (default 1000)

to see if you can get the permissions and ownership in rclone nfsmount to match those in fuse-t. You could try `--umask 0000 --file-perms 777 --dir-perms 777` and that should give all the permissions!

Thank you

**Comment by kapitainsky** [comment]
nfsmount:

```
$ ls -l mount
total 27
-rw-r--r--  1 kptsky  staff  13533 Aug 27 17:44 test.docx

$ ls -ld mount
drwxr-xr-x  1 kptsky  staff  0 Aug 28 19:55 mount

$ stat mount
436207726 18446744073709551615 drwxr-xr-x 1 kptsky staff 0 0 "Jul 14 03:04:00 2042" "Aug 28 19:55:33 2024" "Jul 14 03:04:00 2042" "Jan  1 01:00:00 1970" 1048576 0 0 mount

$ stat mount/test.docx
436207726 18446744073709551615 -rw-r--r-- 1 kptsky staff 0 13533 "Jul 14 03:04:00 2042" "Aug 27 17:44:47 2024" "Jul 14 03:04:00 2042" "Jan  1 01:00:00 1970" 1048576 27 0 mount/test.docx
```

fuse-t:
```
$ ls -l mount
total 27
-rw-r--r--  1 kptsky  staff  13533 Aug 27 17:44 test.docx

$ ls -ld mount
drwxr-xr-x  1 kptsky  staff  0 Aug 28 19:58 mount

$ stat mount
436207727 1 drwxr-xr-x 1 kptsky staff 0 0 "Aug 28 19:58:32 2024" "Aug 28 19:58:32 2024" "Aug 28 19:58:32 2024" "Aug 28 19:58:32 2024" 1048576 0 0 mount

$ stat mount/test.docx
436207727 2 -rw-r--r-- 1 kptsky staff 0 13533 "Aug 27 17:44:47 2024" "Aug 27 17:44:47 2024" "Aug 27 17:44:47 2024" "Aug 27 17:44:47 2024" 1048576 27 0 mount/test.docx
```

the same....


I am afraid it is more complex than simple permissions as for example LibreOffice works without any issues with nfsmount. It is specific to MS Office - like they require fs to support some specific lock etc. Also see https://github.com/willscott/go-nfs/issues/54 - NFS get crazy creating some strange directories etc.

**Comment by kapitainsky** [comment]
When I try with `-o locallocks` I can open and save documents but  this is how it looks after:

```
$ ls -l mount
total 27
-rw-r--r--@ 1 kptsky  staff  13601 Aug 28 20:06 test.docx
drwxr-xr-x@ 1 kptsky  staff      0 Aug 28 20:06 test.docx.sb-a308ef70-2NoZmO
drwxr-xr-x@ 1 kptsky  staff      0 Aug 28 20:06 test.docx.sb-a308ef70-3ypGPP
drwxr-xr-x@ 1 kptsky  staff      0 Aug 28 20:06 test.docx.sb-a308ef70-buTkcS
drwxr-xr-x@ 1 kptsky  staff      0 Aug 28 20:06 test.docx.sb-a308ef70-kVONuI
```

all these dirs are empty. 

I can open and write this doc again - it creates more and more these dirs.

**Comment by kapitainsky** [comment]
tl;dr - `nfsmount` as it is today is still in BETA and have limited use for macOS

How I concluded about...

I was wondering whether macOS NFS client has some problems so I did simple test. I have configured basic NFS export on Debian machine:

```
/home/kptsky/Temp 172.27.150.7(rw,sync,no_subtree_check)
```

Debian 12 supports both NFS3 and NFS4 out of the box:

```
$ sudo cat /proc/fs/nfsd/versions 
+3 +4 +4.1 +4.2
```

so I tried on macOS very basic:

```
$ sudo mount -t nfs -o resvport,rw 172.27.150.12:/home/kptsky/Temp /Users/kptsky/Temp/mount

and

$ sudo mount -t nfs -o resvport,rw -o nfsvers=3 172.27.150.12:/home/kptsky/Temp /Users/kptsky/Temp/mount
```

In both cases all worked without any issues - I could open/edit/write MS Office documents without any hiccups and weird artefacts left behind.

So it is unfortunately this golang implementation still missing some NFS wizardry. I can see that even as it is now it can be useful for some people for some experiments but for sure it is not as [docs](https://rclone.org/commands/rclone_serve_nfs/) states (it should be deleted IMO):

> The primary purpose for this command is to enable [mount command](https://rclone.org/commands/rclone_mount/) on recent macOS versions where installing FUSE is very cumbersome.

FUSE-T is preferred method as it works:). It ticks all boxes of what nfsmount tries unsuccessfully to resolve.

Rclone was early adopter of FUSE-T support - initially it had its problems but by now is mature and all obvious bugs were ironed out. It does not require any kernel extensions and even can be installed and used without `sudo` access.

I wish one day golang implementation gets all rough edges sorted and adds fully working NFS support to rclone,  but for time being it is work in progress IMO.

**Comment by ncw** [comment]
I have merged the --uid/--gid fixes to master for release in v1.68

I think we've come to a dead end for the moment on this...

@kapitainsky maybe we should document what you wrote in the `nfsmount` docs?

**Comment by kapitainsky** [comment]
it is not a dead end but work in progress:) 

NFS is not the easiest thing to implement I am sure - it is even not easy thing to configure properly on systems where it is supported.

It is in beta so some rough edges are expected. I will think about docs update so users are aware of this version shortcomings and keep researching the subject. I am sure that there is a solution - we just do not know yet all details.

Sometimes it is good idea to forget about it for some time - and then eureka might come.

**Comment by ncw** [comment]
> Sometimes it is good idea to forget about it for some time - and then eureka might come.

:-) I felt we were quite close with `-o locallocks` so that might be worth a mention in the docs.

But yes, NFS is complicated... On the other hand people are using `rclone serve nfs` as part of lots of interesting workflows, not just as part of `rclone nfsmount` so despite its rough edges it is getting quite a bit of use.

**Comment by kapitainsky** [comment]
From fuse-t and restic issue:

https://github.com/macos-fuse-t/fuse-t/issues/61#issuecomment-2335104925

It gives some hint re macOS NFS behaviour. I wonder if it can have something to do with issues faces by golang NFS implementation when used with macOS

**Comment by ncw** [comment]
I don't think we have that problem. The nfs library has a file handle cache which solves it. We recently implemented an on disk file handle cache too for getting the file handles to persist between revive restarts. It is a real problem though as rclone files don't have an inode number so we have to make one up. The symptoms when it goes wrong are usually stale NFS handle errors.

(NFS calls the unique IDs for files "file handles". These are usually made from the inode number (another unique id for each file on the disk filing system) but they don't have to be)
